subsys/mpsl/fem/nrf21540_gpio_spi/mpsl_fem_nrf21540_gpio_spi.c

```diff
@@ -0,0 +1,485 @@
+		return -ENOMEM;
+	}
+#endif
+
+#if DT_NODE_HAS_PROP(DT_NODELABEL(nrf_radio_fem), pdn_gpios)
+	uint8_t pdn_gpiote_channel;
+
+	if (nrfx_gpiote_channel_alloc(&pdn_gpiote_channel) != NRFX_SUCCESS) {
+		return -ENOMEM;
+	}
+#endif
+
+	mpsl_fem_nrf21540_gpio_spi_interface_config_t cfg = {
+		.fem_config = {
+			.pa_time_gap_us  =
+				DT_PROP(DT_NODELABEL(nrf_radio_fem),
+					tx_en_settle_time_us),
+			.lna_time_gap_us =
+				DT_PROP(DT_NODELABEL(nrf_radio_fem),
+					rx_en_settle_time_us),
+			.pdn_settle_us   =
+				DT_PROP(DT_NODELABEL(nrf_radio_fem),
+					pdn_settle_time_us),
+			.trx_hold_us     =
+				DT_PROP(DT_NODELABEL(nrf_radio_fem),
+					trx_hold_time_us),
+			.pa_gain_db =
+				CONFIG_MPSL_FEM_NRF21540_TX_GAIN_DB,
+			.pa_gains_db = {
+				[0] = CONFIG_MPSL_FEM_NRF21540_TX_GAIN_DB_POUTA,
+				[1] = CONFIG_MPSL_FEM_NRF21540_TX_GAIN_DB_POUTB,
+			},
+			.lna_gain_db     =
+				CONFIG_MPSL_FEM_NRF21540_RX_GAIN_DB
+		},
+		.pa_pin_config = {
+#if DT_NODE_HAS_PROP(DT_NODELABEL(nrf_radio_fem), tx_en_gpios)
+			.cfg_type      = MPSL_FEM_PIN_CFG_TYPE_PIN_PORT,
+			.enable        = true,
+			.active_high   = MPSL_FEM_GPIO_POLARITY_GET(tx_en_gpios),
+			.p_port        = MPSL_FEM_GPIO_PORT_REG(tx_en_gpios),
+			.gpio_port_no  = MPSL_FEM_GPIO_PORT_NO(tx_en_gpios),
+			.gpio_port_pin = MPSL_FEM_GPIO_PIN_NO(tx_en_gpios),
+			.gpiote_ch_id  = txen_gpiote_channel
+#else
+			MPSL_FEM_DISABLED_GPIOTE_PIN_CONFIG_INIT
+#endif
+		},
+		.lna_pin_config = {
+#if DT_NODE_HAS_PROP(DT_NODELABEL(nrf_radio_fem), rx_en_gpios)
+			.cfg_type      = MPSL_FEM_PIN_CFG_TYPE_PIN_PORT,
+			.enable        = true,
+			.active_high   = MPSL_FEM_GPIO_POLARITY_GET(rx_en_gpios),
+			.p_port        = MPSL_FEM_GPIO_PORT_REG(rx_en_gpios),
+			.gpio_port_no  = MPSL_FEM_GPIO_PORT_NO(rx_en_gpios),
+			.gpio_port_pin = MPSL_FEM_GPIO_PIN_NO(rx_en_gpios),
+			.gpiote_ch_id  = rxen_gpiote_channel
+#else
+			MPSL_FEM_DISABLED_GPIOTE_PIN_CONFIG_INIT
+#endif
+		},
+		.pdn_pin_config = {
+#if DT_NODE_HAS_PROP(DT_NODELABEL(nrf_radio_fem), pdn_gpios)
+			.cfg_type      = MPSL_FEM_PIN_CFG_TYPE_PIN_PORT,
+			.enable        = true,
+			.active_high   = MPSL_FEM_GPIO_POLARITY_GET(pdn_gpios),
+			.p_port        = MPSL_FEM_GPIO_PORT_REG(pdn_gpios),
+			.gpio_port_no  = MPSL_FEM_GPIO_PORT_NO(pdn_gpios),
+			.gpio_port_pin = MPSL_FEM_GPIO_PIN_NO(pdn_gpios),
+			.gpiote_ch_id  = pdn_gpiote_channel
+#else
+			MPSL_FEM_DISABLED_GPIOTE_PIN_CONFIG_INIT
+#endif
+		},
+		.mode_pin_config = {
+#if DT_NODE_HAS_PROP(DT_NODELABEL(nrf_radio_fem), mode_gpios) && \
+	IS_ENABLED(CONFIG_MPSL_FEM_NRF21540_RUNTIME_PA_GAIN_CONTROL)
+			.cfg_type      = MPSL_FEM_PIN_CFG_TYPE_PIN_PORT,
+			.enable        = true,
+			.active_high   = MPSL_FEM_GPIO_POLARITY_GET(mode_gpios),
+			.p_port        = MPSL_FEM_GPIO_PORT_REG(mode_gpios),
+			.gpio_port_no  = MPSL_FEM_GPIO_PORT_NO(mode_gpios),
+			.gpio_port_pin = MPSL_FEM_GPIO_PIN_NO(mode_gpios)
+#else
+			MPSL_FEM_DISABLED_GPIO_CONFIG_INIT
+#endif
+		},
```

Comment:
Well, this was actually originally the way you suggested, it was changed due to this remark from @jciupis : https://github.com/nrfconnect/sdk-nrf/pull/8073#discussion_r914539599

I thing shortening the function as it was done is a good idea, as if we add an SPI-only implementation in the future it would be a bit quicker to create.

Also `mpsl_fem_nrf21540_gpio_spi_interface_config_t` cannot be changed to `const`. In the normal situation (so when using pinctrl) there is no way to check the SPI pins at built time (because they are hidden under the pinctrl). They must aquired by looping through the pins in a pinctrl state, see https://github.com/ahasztag/sdk-nrf/blob/e435b97f5d298faa5426509a190a5a193c300212/subsys/mpsl/fem/nrf21540_gpio_spi/mpsl_fem_nrf21540_gpio_spi.c#L188

Reply:
ok